### PR TITLE
Iss8 tell about you

### PR DESCRIPTION
**Pull request by samsoncsyu7777**
background page: tell about your background.

### CODE REVIEW

File: client/src/App.js [hunk 1]
```diff
@@ -1,19 +1,80 @@
...
+        .then(handleErrors)
+        .then((response) => response.json())
+        .then((data) => {
+          let user = data;
+          setUser(user);
+          console.log(data);
+          if (!user.backgroundCompleted) {
+            return <Redirect to="/background" />;
+          } else {
+            return <Redirect to="/dashboard" />;
```
Comment (skassam21) [inline client/src/App.js]: Does this actually work? I think these returns need to be in the body of the functional component for it to work

Comment (skassam21) [reply client/src/App.js]: Instead, you should do these checks in the body of the functional component

File: client/src/App.js [hunk 2]
```diff
@@ -1,19 +1,80 @@
...
+function App(props) {
+  const [user, setUser] = useState(null);
+
+  const value = useMemo(
+    () => ({
+      user: user,
+      setUser: setUser,
+    }),
+    [user, setUser]
+  );
```
Comment (skassam21) [inline client/src/App.js]: Why is this showing as new code? I thought we merged @abohakobyan pull request (PR)?

File: client/src/index.css [hunk 3]
```diff
@@ -1,3 +1,5 @@
+@import url("https://fonts.googleapis.com/css?family=Open+Sans");
```
Comment (skassam21) [inline client/src/index.css]: Does this do anything? if we add it to index.html, we don't also need this

File: client/src/pages/Background.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+  const handleClick = () => {
+    //the method to handle "next step" button being clicked
+    if (experience === "") {
+      setOpenSnack(true);
+    } else {
+      user.backgroundCompleted = true;
+      user.background.language = language;
+      user.background.experience = experience;
+      user.background.rating = value;
+      setUser(user);
```
Comment (skassam21) [inline client/src/pages/Background.js]: You probably only want to set the user after you get a response from the API and it is successful

File: client/src/pages/Background.js [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        headers: { "Content-Type": "application/json" },
+        body: JSON.stringify(user),
+      })
+        .then((response) => response.json())
+        .then((responseJson) => {
+          console.log(responseJson);
+          return responseJson;
+        })
+        .catch((error) => {
+          console.error(error);
```
Comment (skassam21) [inline client/src/pages/Background.js]: Did you want to show an error message using the snackbar if there is a failure for some reason?

File: client/src/pages/Background.js [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+      user.background.language = language;
+      user.background.experience = experience;
+      user.background.rating = value;
+      setUser(user);
+      console.log(user.background);
+      //fetch here
+      const res = fetch("/" + user.id, {
+        method: "put",
+        headers: { "Content-Type": "application/json" },
+        body: JSON.stringify(user),
```
Comment (skassam21) [inline client/src/pages/Background.js]: The back-end route for this API needs fixing - it right now lets any user edit any user, and needs some authentication

**Review by skassam21 — COMMENTED** [review]
I'm wondering why there is a lot of the changes from other people's code in this pull request?

File: client/src/App.js (hunk 2 above)
Comment (samsoncsyu7777) [reply client/src/App.js]: I need his code for testing my code.

Comment (samsoncsyu7777) [reply client/src/App.js]: should I remove this code first?

File: client/src/pages/Background.js (hunk 4 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: OK, I move it to several lines lower

File: client/src/pages/Background.js (hunk 5 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: Good idea. Let us show the error with the snackbar

File: client/src/pages/Background.js (hunk 6 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: I try to fix at the back end

File: client/src/App.js (hunk 1 above)
Comment (samsoncsyu7777) [reply client/src/App.js]: When the browser is refreshed, I tested it goes through App.js

File: client/src/index.css (hunk 3 above)
Comment (samsoncsyu7777) [reply client/src/index.css]: I try to add to index.html instead

Comment (samsoncsyu7777) [reply client/src/index.css]: Done

File: client/src/pages/Background.js (hunk 4 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: I added {msg: "updated"} as response of success. When detected, setUser and route to dashboard.

File: client/src/pages/Background.js (hunk 5 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: .catch((error) => {
          console.error(error);
          setMessage("Database connection problem!");
          setOpenSnack(true);
        });

File: client/src/pages/Background.js (hunk 6 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: May I fix with this?
router.put("/:id", verifyToken, update); in server/App.js

File: client/src/App.js (hunk 1 above)
Comment (samsoncsyu7777) [reply client/src/App.js]: if (user === null) {
    return <p>Loading profile...</p>;
  } else if (user === "failed to fetch") {
    return <Redirect to="/login" />;
  } else if (!user.backgroundCompleted) {
    return <Redirect to="/background" />;
  }

It works with this logic

**Comment by samsoncsyu7777** [comment]
use Material-UI Responsive full-screen Dialog

File: client/public/index.html [hunk 7]
```diff
@@ -14,6 +14,7 @@
     <link
       rel="stylesheet"
       type="text/css"
+      href="https://fonts.googleapis.com/css?family=Open+Sans"
       href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700|Material+Icons"
```
Comment (skassam21) [inline client/public/index.html]: Only have one href and should be:
`https://fonts.googleapis.com/css?family=Open+Sans|Material+Icons`

File: client/src/pages/Background.js (hunk 5 above)
Comment (skassam21) [reply client/src/pages/Background.js]: Good, remove the commented out code though please

File: client/src/pages/navbar.js [hunk 8]
```diff
@@ -47,7 +47,10 @@ function Navbar(props) {
     return <p>Loading profile...</p>;
   } else if (user === "failed to fetch") {
     return <Redirect to="/login" />;
+  } else if (!user.backgroundCompleted) {
+    return <Redirect to="/background" />;
```
Comment (skassam21) [inline client/src/pages/navbar.js:51]: 👍

File: server/users/users.controller.js [hunk 9]
```diff
@@ -98,10 +98,12 @@ function getById(req, res, next) {
 }
 
 function update(req, res, next) {
-  userService
-    .update(req.params.id, req.body)
-    .then(() => res.json({}))
-    .catch((err) => next(err));
+  if (req.user.id === req.params.id){
+    userService
+      .update(req.params.id, req.body)
```
Comment (skassam21) [inline server/users/users.controller.js:103]: Okay, but you must also check that req.user.id === req.params.id

File: client/public/index.html (hunk 7 above)
Comment (samsoncsyu7777) [reply client/public/index.html]: OK

File: client/src/pages/Background.js (hunk 5 above)
Comment (samsoncsyu7777) [reply client/src/pages/Background.js]: removed

**Pull request merged**